### PR TITLE
listener: create generic listener filter timeout

### PR DESCRIPTION
**Pull request by mattklein123**
Move the hard-coded 15s timeout in TLS inspector into
the connection handler such that it covers all listener
filters. Also make it configurable as well as add useful
stats to see how many connections are currently undergoing
listener filter processing.

Fixes https://github.com/envoyproxy/envoy/issues/5217

*Risk Level*: Medium
*Testing*: Unit tests
*Docs Changes*: Done
*Release Notes*: Done

### CODE REVIEW

File: include/envoy/network/listener.h [hunk 1]
```diff
@@ -58,6 +58,13 @@ class ListenerConfig {
    */
   virtual uint32_t perConnectionBufferLimitBytes() PURE;
 
+  /**
+   * @return std::chrono::milliseconds the timeout to wait for all listener filters to complete
+   *         operation. If the timeout is reached, the accepted socket is closed without a
+   *         connection being created. 0 Specifies a disabled timeout.
```
Comment (venilnoronha) [inline include/envoy/network/listener.h]: nit: Specifies -> specifies

**Comment by mattklein123** [comment]
@ggreenway @PiotrSikora PTAL

File: test/server/listener_manager_impl_test.cc [hunk 2]
```diff
@@ -375,6 +377,22 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, StatsScopeTest) {
   EXPECT_EQ(1UL, server_.stats_store_.counter("listener.127.0.0.1_1234.foo").value());
 }
 
+TEST_F(ListenerManagerImplTest, NotDefaultListerFiltersTimeout) {
```
Comment (venilnoronha) [inline test/server/listener_manager_impl_test.cc]: nit: NotDefaultListerFiltersTimeout -> NotDefaultListenerFiltersTimeout

File: docs/root/intro/version_history.rst [hunk 3]
```diff
@@ -30,6 +30,10 @@ Version history
   value to override the default HTTP to gRPC status mapping.
 * http: no longer close the TCP connection when a HTTP/1 request is retried due
   to a response with empty body.
+* listeners: all listener filters are now governed by the :ref:`listener_filters_timeout
+  <envoy_api_field_Listener.listener_filters_timeout>` setting. The hard coded 15s timeout in
+  the :ref:`TLS inspector listener filter <config_listener_filters_tls_inspector>` is superseded by
```
Comment (alyssawilk) [inline docs/root/intro/version_history.rst]: The prior TLS inspector listener filter _timeout_ is replaced, not the TLS inspector listener, right?

File: include/envoy/network/listener.h [hunk 4]
```diff
@@ -58,6 +58,13 @@ class ListenerConfig {
    */
   virtual uint32_t perConnectionBufferLimitBytes() PURE;
 
+  /**
+   * @return std::chrono::milliseconds the timeout to wait for all listener filters to complete
```
Comment (alyssawilk) [inline include/envoy/network/listener.h]: super nitty: "the timeout to wait for" -> "the timeout for" or "the time to wait for"?

File: source/extensions/filters/listener/tls_inspector/tls_inspector.h [hunk 5]
```diff
@@ -23,7 +23,6 @@ namespace TlsInspector {
   COUNTER(connection_closed)                                                                       \
   COUNTER(client_hello_too_large)                                                                  \
   COUNTER(read_error)                                                                              \
-  COUNTER(read_timeout)                                                                            \
```
Comment (alyssawilk) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.h]: Looks like there aren't docs to update here.  Non-blocking but maybe file a bug against Greg or Piotr to doc up the stats we have?

File: docs/root/intro/version_history.rst (hunk 3 above)
Comment (mattklein123) [reply docs/root/intro/version_history.rst]: It reads: "The hard coded 15s timeout in the TLS inspector listener filter is superseded by..." I think that is OK?

File: source/extensions/filters/listener/tls_inspector/tls_inspector.h (hunk 5 above)
Comment (mattklein123) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.h]: https://github.com/envoyproxy/envoy/issues/5235

**Pull request merged**